### ISSUE TITLE
Memory leak in Perl_yylex (toke.c:7027)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15829]
Migrated from [rt.perl.org#130632](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=130632) (status was 'resolved')

Searchable as RT130632$

**Comment by p5pRT** [comment]
### From @dur-randir

#### Created by @dur-randir



While fuzzing perl v5\.25\.8\-216\-gfbceb79751 built with afl and run
under libdislocator\, I found the following program

s//'x' ^ \-"s&#8203;:\\347"/eeg

to cause a memory leak report under ASAN&#8203;:

=================================================================
==27209==ERROR&#8203;: LeakSanitizer&#8203;: detected memory leaks

Direct leak of 4 byte\(s\) in 1 object\(s\) allocated from&#8203;:
&nbsp;    \#0 0x4ea6c8 in malloc \(/home/afl/afl\-asan/perl\+0x4ea6c8\)
&nbsp;    \#1 0x8504be in Perl\_safesysmalloc /home/afl/afl\-asan/util\.c&#8203;:153&#8203;:21
&nbsp;    \#2 0x8564ab in Perl\_savepvn /home/afl/afl\-asan/util\.c&#8203;:1177&#8203;:5
&nbsp;    \#3 0x675843 in Perl\_yylex /home/afl/afl\-asan/toke\.c&#8203;:7027&#8203;:23
&nbsp;    \#4 0x6f6072 in Perl\_yyparse /home/afl/afl\-asan/perly\.c&#8203;:340&#8203;:34
&nbsp;    \#5 0xb0c78f in S\_doeval\_compile /home/afl/afl\-asan/pp\_ctl\.c&#8203;:3432&#8203;:77
&nbsp;    \#6 0xb09a17 in Perl\_pp\_entereval /home/afl/afl\-asan/pp\_ctl\.c&#8203;:4292&#8203;:9
&nbsp;    \#7 0x84b114 in Perl\_runops\_debug /home/afl/afl\-asan/dump\.c&#8203;:2406&#8203;:23
&nbsp;    \#8 0x5f1b05 in S\_run\_body /home/afl/afl\-asan/perl\.c&#8203;:2528&#8203;:2
&nbsp;    \#9 0x5f1b05 in perl\_run /home/afl/afl\-asan/perl\.c&#8203;:2451
&nbsp;    \#10 0x5224d2 in main /home/afl/afl\-asan/perlmain\.c&#8203;:123&#8203;:9
&nbsp;    \#11 0x7efca17d02b0 in \_\_libc\_start\_main
\(/lib/x86\_64\-linux\-gnu/libc\.so\.6\+0x202b0\)

SUMMARY&#8203;: AddressSanitizer&#8203;: 4 byte\(s\) leaked in 1 allocation\(s\)\.

Leaked scalar comes from the following line

pl\_yylval\.pval = savepvn\(PL\_tokenbuf\, len\+1\);

In this case PL\_tokenbuf is "Us" and len == 2\. Amount of memory leaked
scales with the size of this buffer\. For example\, the following
eternal loop leaks 20 bytes per iteration\, so run it with caution&#8203;:

s//'x' ^ \-"s666666666666666k&#8203;:\\347"/eeg while \(1\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.25.9:

Configured by root at Sat Jan 14 02:25:05 MSK 2017.

Summary of my perl5 (revision 5 version 25 subversion 9) configuration:
  Commit id: cbe2fc5001aa59cdc73e04cc35e097a2ecfbeec0
  Platform:
    osname=linux
    osvers=3.16.0-4-amd64
    archname=x86_64-linux
    uname='linux dorothy 3.16.0-4-amd64 #1 smp debian 3.16.36-1+deb8u2
(2016-10-19) x86_64 gnulinux '
    config_args='-des -Dusedevel -DDEBUGGING -Dcc=afl-clang-fast
-Doptimize=-O0 -g -ggdb3'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    bincompat5005=undef
  Compiler:
    cc='afl-clang-fast'
    ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe
-fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE
-D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O0 -g -ggdb3'
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe
-fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='4.2.1 Compatible Clang 3.9.1 (tags/RELEASE_391/rc2)'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='afl-clang-fast'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/llvm-3.9/bin/../lib/clang/3.9.1/lib
/usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu
/lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.24.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.24'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O0 -g -ggdb3 -L/usr/local/lib -fstack-protector-strong'



@INC for perl 5.25.9:
    lib
    /usr/local/lib/perl5/site_perl/5.25.9/x86_64-linux
    /usr/local/lib/perl5/site_perl/5.25.9
    /usr/local/lib/perl5/5.25.9/x86_64-linux
    /usr/local/lib/perl5/5.25.9


Environment for perl 5.25.9:
    HOME=/home/afl
    LANG=en_US.UTF-8
    LANGUAGE=en_US:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/afl/perlbrew/bin:/home/afl/perlbrew/perls/perl-5.22.1/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
    PERLBREW_BASHRC_VERSION=0.78
    PERLBREW_HOME=/home/afl/.perlbrew
    PERLBREW_MANPATH=/home/afl/perlbrew/perls/perl-5.22.1/man
    PERLBREW_PATH=/home/afl/perlbrew/bin:/home/afl/perlbrew/perls/perl-5.22.1/bin
    PERLBREW_PERL=perl-5.22.1
    PERLBREW_ROOT=/home/afl/perlbrew
    PERLBREW_VERSION=0.78
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue\, 24 Jan 2017 00&#8203;:19&#8203;:10 \-0800\, randir wrote&#8203;:
> While fuzzing perl v5\.25\.8\-216\-gfbceb79751 built with afl and run
> under libdislocator\, I found the following program
> 
> s//'x' ^ \-"s&#8203;:\\347"/eeg
> 
> to cause a memory leak report under ASAN&#8203;:
> 
> =================================================================
> ==27209==ERROR&#8203;: LeakSanitizer&#8203;: detected memory leaks
> 
> Direct leak of 4 byte\(s\) in 1 object\(s\) allocated from&#8203;:
>     \#0 0x4ea6c8 in malloc \(/home/afl/afl\-asan/perl\+0x4ea6c8\)
>     \#1 0x8504be in Perl\_safesysmalloc /home/afl/afl\-asan/util\.c&#8203;:153&#8203;:21
>     \#2 0x8564ab in Perl\_savepvn /home/afl/afl\-asan/util\.c&#8203;:1177&#8203;:5
>     \#3 0x675843 in Perl\_yylex /home/afl/afl\-asan/toke\.c&#8203;:7027&#8203;:23
>     \#4 0x6f6072 in Perl\_yyparse /home/afl/afl\-asan/perly\.c&#8203;:340&#8203;:34
>     \#5 0xb0c78f in S\_doeval\_compile /home/afl/afl\-
> asan/pp\_ctl\.c&#8203;:3432&#8203;:77
>     \#6 0xb09a17 in Perl\_pp\_entereval /home/afl/afl\-
> asan/pp\_ctl\.c&#8203;:4292&#8203;:9
>     \#7 0x84b114 in Perl\_runops\_debug /home/afl/afl\-asan/dump\.c&#8203;:2406&#8203;:23
>     \#8 0x5f1b05 in S\_run\_body /home/afl/afl\-asan/perl\.c&#8203;:2528&#8203;:2
>     \#9 0x5f1b05 in perl\_run /home/afl/afl\-asan/perl\.c&#8203;:2451
>     \#10 0x5224d2 in main /home/afl/afl\-asan/perlmain\.c&#8203;:123&#8203;:9
>     \#11 0x7efca17d02b0 in \_\_libc\_start\_main
> \(/lib/x86\_64\-linux\-gnu/libc\.so\.6\+0x202b0\)

This normally freed by newSTATEOP\(\)\, but if the label token happens to have been pushed onto the parser shift\-reduce stack and the parser aborts\, it leaks\.

The attached fixes it for me\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-perl-130632-free-label-pvs-from-the-parser-stack.patch</summary>

```diff
From 7cc5dfff56d7dfaac8a7f7cbb6c2ff96f3036eb4 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Tue, 7 Feb 2017 15:08:55 +1100
Subject: (perl #130632) free label pvs from the parser stack

For labels, Perl_yylex() generates a LABEL token, with the label name
stored in the pval slot of the yylval, allocated with savepvn().

In the normal course of parsing this is freed by Perl_newSTATEOP(), but
if parsing is aborted with the label on the parser shift-reduce stack
the memory would leak.

Clean up pval entries on the parse stack when clearing the parser
stack.
---
 perly.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/perly.c b/perly.c
index 1c018bb..43cc955 100644
--- a/perly.c
+++ b/perly.c
@@ -232,6 +232,10 @@ S_clear_yystack(pTHX_  const yy_parser *parser)
 	    YYDPRINTF ((Perl_debug_log, "(freeing op)\n"));
 	    op_free(ps->val.opval);
 	}
+        else if (yy_type_tab[yystos[ps->state]] == toketype_pval
+                 && ps->val.pval) {
+            Safefree(ps->val.pval);
+        }
 	SvREFCNT_dec(ps->compcv);
 	ps--;
     }
-- 
2.1.4


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue\, 24 Jan 2017 00&#8203;:19&#8203;:10 \-0800\, randir wrote&#8203;:
> While fuzzing perl v5\.25\.8\-216\-gfbceb79751 built with afl and run
> under libdislocator\, I found the following program
> 
> s//'x' ^ \-"s&#8203;:\\347"/eeg
> 
> to cause a memory leak report under ASAN&#8203;:
> 
> =================================================================
> ==27209==ERROR&#8203;: LeakSanitizer&#8203;: detected memory leaks
> 
> Direct leak of 4 byte\(s\) in 1 object\(s\) allocated from&#8203;:
>     \#0 0x4ea6c8 in malloc \(/home/afl/afl\-asan/perl\+0x4ea6c8\)
>     \#1 0x8504be in Perl\_safesysmalloc /home/afl/afl\-asan/util\.c&#8203;:153&#8203;:21
>     \#2 0x8564ab in Perl\_savepvn /home/afl/afl\-asan/util\.c&#8203;:1177&#8203;:5
>     \#3 0x675843 in Perl\_yylex /home/afl/afl\-asan/toke\.c&#8203;:7027&#8203;:23
>     \#4 0x6f6072 in Perl\_yyparse /home/afl/afl\-asan/perly\.c&#8203;:340&#8203;:34
>     \#5 0xb0c78f in S\_doeval\_compile /home/afl/afl\-
> asan/pp\_ctl\.c&#8203;:3432&#8203;:77
>     \#6 0xb09a17 in Perl\_pp\_entereval /home/afl/afl\-
> asan/pp\_ctl\.c&#8203;:4292&#8203;:9
>     \#7 0x84b114 in Perl\_runops\_debug /home/afl/afl\-asan/dump\.c&#8203;:2406&#8203;:23
>     \#8 0x5f1b05 in S\_run\_body /home/afl/afl\-asan/perl\.c&#8203;:2528&#8203;:2
>     \#9 0x5f1b05 in perl\_run /home/afl/afl\-asan/perl\.c&#8203;:2451
>     \#10 0x5224d2 in main /home/afl/afl\-asan/perlmain\.c&#8203;:123&#8203;:9
>     \#11 0x7efca17d02b0 in \_\_libc\_start\_main
> \(/lib/x86\_64\-linux\-gnu/libc\.so\.6\+0x202b0\)
> 
> SUMMARY&#8203;: AddressSanitizer&#8203;: 4 byte\(s\) leaked in 1 allocation\(s\)\.
> 
> Leaked scalar comes from the following line
> 
> pl\_yylval\.pval = savepvn\(PL\_tokenbuf\, len\+1\);
> 
> In this case PL\_tokenbuf is "Us" and len == 2\. Amount of memory leaked
> scales with the size of this buffer\. For example\, the following
> eternal loop leaks 20 bytes per iteration\, so run it with caution&#8203;:
> 
> s//'x' ^ \-"s666666666666666k&#8203;:\\347"/eeg while \(1\)

This looks like it's fixed\, I expect by&#8203;:

commit 017192018b7f9fc4b889bb344bd75df8f6b78929
Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
Date&#8203;:   Fri Apr 5 12&#8203;:49&#8203;:50 2019 \+0100

&nbsp;    parser&#8203;: change LABEL type from pval to opval
&nbsp;    
&nbsp;    The items pushed onto the parser stack can be one of several types&#8203;:
&nbsp;    ival\, opval\, pval etc\. The only remaining use of pval is when a "label&#8203;:"
&nbsp;    is encountered\.
&nbsp;    
&nbsp;    When an error occurs during parsing\, ops on the parse stack get
&nbsp;    automatically reaped these days as part of the OP slab mechanism;
&nbsp;    but bare strings \(pvals\) still leak\.
&nbsp;    
&nbsp;    Convert this one remaining pval into an opval\, making the toker return
&nbsp;    an OP\_CONST with an SV holding the label\.
&nbsp;    
&nbsp;    Since newSTATEOP\(\) still expects a raw string for the label\, the parser
&nbsp;    just grabs the value returned by the toker and makes a copy of the
&nbsp;    string from it\, then immediately frees the OP\_CONST and its associated
&nbsp;    SV\.
&nbsp;    
&nbsp;    The leak was showing up in ext/XS\-APItest/t/stmtasexpr\.t\, which expects
&nbsp;    to parse a statement where labels are banned\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue\, 11 Jun 2019 23&#8203;:21&#8203;:07 \-0700\, tonyc wrote&#8203;:
> On Tue\, 24 Jan 2017 00&#8203;:19&#8203;:10 \-0800\, randir wrote&#8203;:
> > While fuzzing perl v5\.25\.8\-216\-gfbceb79751 built with afl and run
> > under libdislocator\, I found the following program
> > 
> > s//'x' ^ \-"s&#8203;:\\347"/eeg
> > 
> > to cause a memory leak report under ASAN&#8203;:
> > 
> > =================================================================
> > ==27209==ERROR&#8203;: LeakSanitizer&#8203;: detected memory leaks
> > 
> > Direct leak of 4 byte\(s\) in 1 object\(s\) allocated from&#8203;:
> >     \#0 0x4ea6c8 in malloc \(/home/afl/afl\-asan/perl\+0x4ea6c8\)
> >     \#1 0x8504be in Perl\_safesysmalloc /home/afl/afl\-asan/util\.c&#8203;:153&#8203;:21
> >     \#2 0x8564ab in Perl\_savepvn /home/afl/afl\-asan/util\.c&#8203;:1177&#8203;:5
> >     \#3 0x675843 in Perl\_yylex /home/afl/afl\-asan/toke\.c&#8203;:7027&#8203;:23
> >     \#4 0x6f6072 in Perl\_yyparse /home/afl/afl\-asan/perly\.c&#8203;:340&#8203;:34
> >     \#5 0xb0c78f in S\_doeval\_compile /home/afl/afl\-
> > asan/pp\_ctl\.c&#8203;:3432&#8203;:77
> >     \#6 0xb09a17 in Perl\_pp\_entereval /home/afl/afl\-
> > asan/pp\_ctl\.c&#8203;:4292&#8203;:9
> >     \#7 0x84b114 in Perl\_runops\_debug /home/afl/afl\-asan/dump\.c&#8203;:2406&#8203;:23
> >     \#8 0x5f1b05 in S\_run\_body /home/afl/afl\-asan/perl\.c&#8203;:2528&#8203;:2
> >     \#9 0x5f1b05 in perl\_run /home/afl/afl\-asan/perl\.c&#8203;:2451
> >     \#10 0x5224d2 in main /home/afl/afl\-asan/perlmain\.c&#8203;:123&#8203;:9
> >     \#11 0x7efca17d02b0 in \_\_libc\_start\_main
> > \(/lib/x86\_64\-linux\-gnu/libc\.so\.6\+0x202b0\)
> > 
> > SUMMARY&#8203;: AddressSanitizer&#8203;: 4 byte\(s\) leaked in 1 allocation\(s\)\.
> > 
> > Leaked scalar comes from the following line
> > 
> > pl\_yylval\.pval = savepvn\(PL\_tokenbuf\, len\+1\);
> > 
> > In this case PL\_tokenbuf is "Us" and len == 2\. Amount of memory leaked
> > scales with the size of this buffer\. For example\, the following
> > eternal loop leaks 20 bytes per iteration\, so run it with caution&#8203;:
> > 
> > s//'x' ^ \-"s666666666666666k&#8203;:\\347"/eeg while \(1\)
> 
> This looks like it's fixed\, I expect by&#8203;:
> 
> commit 017192018b7f9fc4b889bb344bd75df8f6b78929
> Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
> Date&#8203;:   Fri Apr 5 12&#8203;:49&#8203;:50 2019 \+0100
> 
>     parser&#8203;: change LABEL type from pval to opval

And so closing\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
